File: AndroidApp/src/main/java/com/smp/sharedkotlin/app/presentation/MemoListViewModel.kt
package com.smp.sharedkotlin.app.presentation

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import com.smp.sharedkotlin.app.support.*
import com.smp.sharedkotlin.domain.GetAllMemosUseCase


class MemoListViewModel(
    private val schedulerProvider: SchedulerProvider,
    private val getAllMemosUseCase: GetAllMemosUseCase
) : BaseViewModel() {

    private val _memos = MutableLiveData<List<Memo>>()
    val memos: LiveData<List<Memo>> = _memos
    val memosSyncProgress = MediatorLiveData<Progress>()

    fun syncMemos() {
        SingleSupport.deferredJust { getAllMemosUseCase.execute() }
            .subscribeOn(schedulerProvider.io())
            .map(MemoMapper::mapAll)
            .observeOn(schedulerProvider.ui())
            .syncProgress(memosSyncProgress)
            .subscribe { result ->
                _memos.value = result
            }
            .bind(this)

    }
}

File: iOSApp/iOSApp/Presentation/MemoListViewModel.swift
import SharedCode
import RxSwift

class MemoListViewModel : BaseViewModel {
    
    private let schedulerProvider: SchedulerProvider
    private let getAllMemosUseCase: GetAllMemosUseCase
    private let memoMapper = MemoMapper()
    
    private let memoPublisher = ReplaySubject<[Memo]>.create(bufferSize: 1)
    
    init(schedulerProvider: SchedulerProvider, getAllMemosUseCase: GetAllMemosUseCase) {
        self.schedulerProvider = schedulerProvider
        self.getAllMemosUseCase = getAllMemosUseCase
    }
    
    func observeMemos() -> Observable<[Memo]> {
        return memoPublisher
    }
    
    func syncMemos() {
        SingleSupport.deferredJust(fun: { self.getAllMemosUseCase.execute() })
            .subscribeOn(schedulerProvider.io())
            .map { (memos) -> [Memo] in self.memoMapper.mapAll(objs: memos) }
            .observeOn(schedulerProvider.ui())
            .subscribe { event in
                switch event {
                case .success(let memos):
                    self.memoPublisher.onNext(memos)
                case .error(let error):
                    print("Error: ", error)
                }
            }
            .disposed(by: disposeBag)
    }
}
